soroban-env-host/src/host.rs

```diff
@@ -2438,50 +2440,82 @@ impl VmCallerEnv for Host {
+            // Approximate the strkey encoding cost with two vector allocations:
+            // one for the payload size (32-byte key/hash + 3 bytes for
+            // version/checksum) and  another one for the base32 encoding of
+            // the payload.
+            const PAYLOAD_LEN: u64 = 32 + 3;
+            Vec::<u8>::charge_bulk_init_cpy(PAYLOAD_LEN + (PAYLOAD_LEN * 8 + 4) / 5, self)?;
+            let strkey = match addr {
+                ScAddress::Account(acc_id) => {
+                    let AccountId(PublicKey::PublicKeyTypeEd25519(Uint256(ed25519))) = acc_id;
+                    let strkey = stellar_strkey::Strkey::PublicKeyEd25519(
+                        stellar_strkey::ed25519::PublicKey(ed25519.metered_clone(self)?),
+                    );
+                    strkey
+                }
+                ScAddress::Contract(Hash(h)) => stellar_strkey::Strkey::Contract(
+                    stellar_strkey::Contract(h.metered_clone(self)?),
+                ),
+            };
+            Ok(strkey.to_string())
+        })?;
+        self.add_host_object(ScString(strkey.try_into()?))
     }
 
-    fn contract_id_to_address(
+    fn strkey_to_address(
         &self,
         _vmcaller: &mut VmCaller<Self::VmUserState>,
-        contract_id_bytes: BytesObject,
+        strkey_obj: StringObject,
     ) -> Result<AddressObject, Self::Error> {
-        let contract_id = self.hash_from_bytesobj_input("contract_id", contract_id_bytes)?;
-        self.add_host_object(ScAddress::Contract(contract_id))
-    }
-
-    fn address_to_account_public_key(
-        &self,
-        _vmcaller: &mut VmCaller<Self::VmUserState>,
-        address: AddressObject,
-    ) -> Result<Val, Self::Error> {
-        let addr = self.visit_obj(address, |addr: &ScAddress| addr.metered_clone(self))?;
-        match addr {
-            ScAddress::Account(AccountId(PublicKey::PublicKeyTypeEd25519(pk))) => Ok(self
-                .add_host_object(ScBytes(self.metered_slice_to_vec(&pk.0)?.try_into()?))?
-                .into()),
-            ScAddress::Contract(_) => Ok(().into()),
-        }
-    }
+        let sc_addr = self.visit_obj(strkey_obj, |key: &ScString| {
+            const PAYLOAD_LEN: u64 = 32 + 3;
+            let expected_key_len = (PAYLOAD_LEN * 8 + 4) / 5;
+            if expected_key_len != key.0.len() as u64 {
+                return Err(self.err(
+                    ScErrorType::Value,
+                    ScErrorCode::InvalidInput,
+                    "unexpected strkey length",
+                    &[strkey_obj.to_val()],
+                ));
+            }
+            // Approximate the decoding cost as two vector allocations for the
+            // expected payload length (the strkey library does one extra copy).
+            Vec::<u8>::charge_bulk_init_cpy(PAYLOAD_LEN + PAYLOAD_LEN, self)?;
+            // Charge for the key copy to string.
+            // Vec::<u8>::charge_bulk_init_cpy(key.len() as u64, self)?;
```

Comment:
This this line supposed to be uncommented, or removed?

Reply:
Nice catch, I thought I can avoid a copy, but apparently no, so need to charge another copy.